### DISCUSSION TITLE
PreProcessor for XAML preview using DI

### DISCUSSION THREAD

**Discussion by JohnKesko** [AvaloniaUI/Avalonia#12508] [General]
Does anyone have a nice way to handle the XAML previewer in design time vs when I need to inject dependencies?
Since design expects a parameterless constructor, this is how I see it or like to achieve. 
It would be nice to handle it automagically when pressing debug so you can use both constructors by just having preprocessors defined.

I guess it's a combo of Rider and Avalonia, so does anyone have an idea?

```
public partial class MainWindowViewModel : ViewModelBase
{
    private readonly INavigationService _navigationService;

#if DESIGN
    public MainWindowViewModel()
    {
        CurrentPage = new LoginPageViewModel();
    }
#endif
    
    public MainWindowViewModel(INavigationService navigationService)
    {
        _navigationService = navigationService;
        
        CurrentPage = _navigationService.CurrentPage;
        
        // Subscribe to events
        _navigationService.NavigationOccurred += OnNavigationOccurred;
        ResizeRequested += OnResizeRequested;
    }

```

**Comment by Mrxx99** [comment]
What I did was inheriting a DesignTimeViewModel from my normal ViewModel, add a parameterless construcor that can either pass in nulls if supported or dummy values. You could also create a protected parameterless construcor in your ViewModel and inherit that in the DesignTimeViewModel. Than in View for Design.DataContext I reference the DesignTimeViewModel

**Reply by JohnKesko** [reply]
Ah, I see. That's pretty neat. Thanks.

**Comment by thevortexcloud** [comment]
As far as I am aware there is no way to do a preprocessor like this. As the way the previewer works is by actually running your app with some special presentation logic (you can easily see that by throwing an exception in your constructor, which should cause the previewer to crash with the error). What you can do is detect at runtime if you are in design mode and then try to do different logic then. But I am not sure that will actually help you here. EG

```csharp
if (Avalonia.Controls.Design.IsDesignMode) {
    DoDesignTimeStuff();
}
else {
    DoRealStuff();
}

```

In theory this should allow you to do dependency injection with dummy services, but that does not seem to work properly. It almost works but then the designer in Rider does not actually display anything, despite it locating all the services. 

The other option is you can pass design time values into the constructor via the XAML. But it's a fairly verbose thing to do, and still needs the default constructor.

I personally just ignore the previewer and make my constructors do what ever and stare at the XAML. It works fine at run time.

**Reply by timunie** [reply]
yeah, that's a valid solution. Or use a speicial Desing-Time instnace

**Reply by JohnKesko** [reply]
Yeah, I see what you mean. Anyway, since #if DESIGN don't exist, I'll just use it as a "comment" for myself. Maybe that's enough if I'm not missing anything obvious.

**Reply by thevortexcloud** [reply]
At that point you may as well just use a region pre-processor directive, that way you don't need to fiddle with build settings.
